README.md
# 🔝**yant** - _**Y**et **A**nother **N**vidia **T**op_ 

![Python 3.5+](https://img.shields.io/badge/Python-3.5%2B-brightgreen.svg)

If there is one thing we don't have enough of, it's the amount of ways to make `nvidia-smi` actually readable. 

Here's ***Y**et **A**nother **N**vidia **T**op* (or `yant`) written in simple, straightforward python.

## Demo

<p align="center">
  <img width="70%" src="demo.png" alt="Monitor">
  </br>
  Sample output of <code>yant</code>.
  </br>
  (Note: <code>USER</code> is removed for privacy protection)
</p>

<p align="center">
  <img width="70%" src="demo-color.png" alt="Monitor">
  </br>
  <code>yant</code> with customized colors.
  </br>
</p>

## Features
- Based on `nvidia-smi` but is more human readable
- Ability to **view the current command** ran on GPUs for process control
- **Colored outputs** with *lots of* options for customization
- No more `NVIDIA GeForce ...`! `yant` gives the option to toggle gpu name display
- ...and many more to come

**If `yant` is useful to you, please star ⭐️ it to let more people know 🤗, or consider contributing to it!**

## Installation

First, install the required `termcolor` package with 

```bash
pip3 install termcolor
``` 

Then, you may clone this repository with 

```bash
git clone https://github.com/Cveinnt/yant.git
```

Next `cd yant` into the directory and use `yant` with 

```bash
python3 yant.py
```

Feel free to use flags to customize the outputs. 

**(Optional)** To make `yant` more `top`-like, use the `watch` utility: `watch -c python3 yant.py`. 

### **Recommended** 
If you want, you can just copy 
```bash
alias yant="watch -c python3 /path/to/yant/yant.py"
```
to your `~/.bashrc` or `~/.zshrc` file, where `/path/to/yant/` is the output of `pwd` when you are in `yant`'s directory.
**Don't forget to `source ~/.zshrc`!** Then you can just run 
```bash
$ yant
```
and watch the magic happens.

To exit `watch`, simply use `Ctrl+C`.

## Usage

    $ python3 yant.py

      Print current GPU utilization with better formatting and color support.

      --length [int]                   Adjust length of printed command used by different 
                                       GPUs. `[int]` should be an integer that represents 
                                       the desired amount of characters to print.
                                       Default: print first 100 characters (int=100).
      
      --high [float]                   Set threshold value for what is considered a high 
                                       GPU usage. Should be a float value between 0 and 1.
                                       Default: 0.8

      --low [float]                    Set threshold value for what is considered a low 
                                       GPU usage. Should be a float value between 0 and 1.
                                       Default: 0.1
      
      --high-color [str]               Customize color output for what is considered a high 
                                       GPU usage. Should be a string.
                                       Default: 'red'

      --mid-color [str]                Customize color output for what is considered a mid-level 
                                       GPU usage. Should be a string.
                                       Default: 'yellow'

      --low-color [str]                Customize color output for what is considered a low 
                                       GPU usage. Should be a string.
                                       Default: 'green'
                                       
      --show-gpu                       Toggle GPU name disply. Sometimes `nvidia-smi` 
                                       does not display the full GPU name. Use this flag to 
                                       display the full name. Only available to GeForce GPUs. 


Outputs are color coded, specifically: 
- red means: `GPU utilization` or `GPU memory` >= `high`
- yellow means: `low` <= `GPU utilization` or `GPU memory` < `high`
- green means: `GPU utilization` or `GPU memory` < `low`

As shown above, you may also change the colors if you'd like. Here's a list of available colors: `["grey", "red", "green", "yellow", "blue", "magenta", "cyan", "white"]`

*Note:* for backward compatibility, `nvidia-smi | python3 yant.py` is also supported.

*Note:* running inside a container (docker, singularity, ...), `nvidia-smi` can only see processes running in the container.


## License
```
DO WHATEVER THE FUCK YOU WANT TO PUBLIC LICENSE 

Copyright (C) 2022 Vincent Wu

Everyone is permitted to copy and distribute verbatim or modified 
copies of this license document, and changing it is allowed as 
long as the name is changed. 

TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION, AND MODIFICATION 

1. DO WHATEVER THE FUCK YOU WANT TO.
2. What did I just say?
```

## Credits
`yant` is inspired by and based heavily on [nvidia-htop](https://github.com/peci1/nvidia-htop).


CHANGELOG.md
# Changelog

## 1.0.1

- First release

yant.py
#!/usr/bin/env python3
import sys
import re
import subprocess
import select
import argparse
from termcolor import colored

colors = ["grey", "red", "green", "yellow", "blue", "magenta", "cyan", "white"]

parser = argparse.ArgumentParser()
parser.add_argument('--length', default=20, const=100, type=int, nargs='?')
parser.add_argument('--high', default=0.8, type=float, nargs='?')
parser.add_argument('--low', default=0.1, type=float, nargs='?')
parser.add_argument('--show-gpu', action='store_true')
#parser.add_argument('--colored', action='store_true') # removed after consideration
parser.add_argument('--high-color', default='red', type=str)
parser.add_argument('--mid-color', default='yellow', type=str)
parser.add_argument('--low-color', default='green', type=str)
args = parser.parse_args()
length, high, low, show = args.length, args.high, args.low, args.show_gpu
highC, midC, lowC = args.high_color, args.mid_color, args.low_color

assert high < 1 and high > 0 and low < 1 and low > 0, "high or low must be within (0,1)!"
assert high > low, "high must be higher than low!"
assert highC in colors and midC in colors and lowC in colors, "Color must be within accepted colors, e.g. {}".format(colors)

MEMORY_MODERATE_RATIO = GPU_MODERATE_RATIO = high
MEMORY_FREE_RATIO = GPU_FREE_RATIO = low

stdin_lines = []
if sys.stdin in select.select([sys.stdin], [], [], 0)[0]:
    stdin_lines = sys.stdin.readlines()
if stdin_lines:
    lines = stdin_lines
else:
    ps_call = subprocess.run('nvidia-smi', stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    ps_name = subprocess.run('nvidia-smi --query-gpu=gpu_name --format=csv,nounits,noheader', stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
    if ps_call.returncode != 0:
        print('nvidia-smi exited with error code {}:'.format(ps_call.returncode))
        print(ps_call.stdout.decode() + ps_call.stderr.decode())
        sys.exit()
    lines_proc = ps_call.stdout.decode().split("\n")
    lines = [line + '\n' for line in lines_proc[:-1]]
    lines += lines_proc[-1]
    gpu_names = ps_name.stdout.decode().split("\n")
    gpus = [gpu + '\n' for gpu in gpu_names[:-1]]
    gpus += gpu_names[-1]


def colorize(_lines):
    for j in range(len(_lines)):
        line = _lines[j]
        m = re.match(r"\| (?:N/A|..%)\s+[0-9]{2,3}C.*\s([0-9]+)MiB\s+/\s+([0-9]+)MiB.*\s([0-9]+)%", line)
        if m is not None:
            used_mem = int(m.group(1))
            total_mem = int(m.group(2))
            gpu_util = int(m.group(3)) / 100.0
            mem_util = used_mem / float(total_mem)
            is_moderate = False
            is_high = gpu_util >= GPU_MODERATE_RATIO or mem_util >= MEMORY_MODERATE_RATIO
            if not is_high:
                is_moderate = gpu_util >= GPU_FREE_RATIO or mem_util >= MEMORY_FREE_RATIO
            c = highC if is_high else (midC if is_moderate else lowC)
            _lines[j] = colored(_lines[j], c)
            _lines[j-1] = colored(_lines[j-1], c)
    return _lines

lines_to_print = []
is_new_format = False

# if showgpu
if show:
    for i in range(len(gpus)):
        lines = [re.sub('{}  NVIDIA GeForce ... '.format(i), '{}  GeForce '.format(i)+gpus[i].rstrip()[15:], l) for l in lines]
# Copy the utilization upper part verbatim
for i in range(len(lines)):
    if not lines[i].startswith("| Processes:"):
        lines_to_print.append(lines[i].rstrip())
    else:
        while not lines[i].startswith("|===="):
            m = re.search(r'GPU\s*GI\s*CI', lines[i])
            if m is not None:
                is_new_format = True
            i += 1
        i += 1
        break

# colorize outputs. absolutely crucial.
#if color:
lines_to_print = colorize(lines_to_print)

# we print all but the last line which is the +---+ separator
for line in lines_to_print[:-1]:
    print(line)

no_running_process = "No running processes found"
if no_running_process in lines[i] or lines[i].startswith("+--"):
    print(lines[-1].strip())
    print("| " + no_running_process + " " * (73 - len(no_running_process)) + "   |")
    if lines[i].startswith("+--"):
        print("| If you're running in a container, you'll only see processes running inside. |")
    print(lines[-1])
    sys.exit()

# Parse the PIDs from the lower part
gpu_num = []
pid = []
gpu_mem = []
user = []
cpu = []
mem = []
time = []
command = []

gpu_num_idx = 1
pid_idx = 2 if not is_new_format else 4
gpu_mem_idx = -3

while not lines[i].startswith("+--"):
    if "Not Supported" in lines[i]:
        i += 1
        continue
    line = lines[i]
    line = re.split(r'\s+', line)
    gpu_num.append(line[gpu_num_idx])
    pid.append(line[pid_idx])
    gpu_mem.append(line[gpu_mem_idx])
    user.append("")
    cpu.append("")
    mem.append("")
    time.append("")
    command.append("")
    i += 1

# Query the PIDs using ps
ps_format = "pid,user,%cpu,%mem,etime,command"
ps_call = subprocess.run(["ps", "-o", ps_format, "-p", ",".join(pid)], stdout=subprocess.PIPE)
processes = ps_call.stdout.decode().split("\n")

# Parse ps output
for line in processes:
    if line.strip().startswith("PID") or len(line) == 0:
        continue
    parts = re.split(r'\s+', line.strip(), 5)
    # idx = pid.index(parts[0])
    for idx in filter(lambda p: pid[p] == parts[0], range(len(pid))):
        user[idx] = parts[1]
        cpu[idx] = parts[2]
        mem[idx] = parts[3]
        time[idx] = parts[4] if "-" not in parts[4] else parts[4].split("-")[0] + " days"
        command[idx] = parts[5]

max_pid_length = max(5, max([len(x) for x in pid]))
format = ("|  %3s %" + str(max_pid_length) + "s %8s   %8s %5s %5s %9s  %-" + str(length) + "." + str(length) + "s  |")

line = format % (
    "GPU", "PID", "USER", "GPU MEM", "%CPU", "%MEM", "TIME", "COMMAND"
)

print("+" + ("-" * (len(line) - 2)) + "+")

print(line)

for i in range(len(pid)):
    print(format % (
        gpu_num[i],
        pid[i],
        user[i],
        gpu_mem[i],
        cpu[i],
        mem[i],
        time[i],
        command[i]
    ))

print("+" + ("-" * (len(line) - 2)) + "+")